tests/test_index.py

```diff
@@ -0,0 +1,118 @@
+
+VALID_INDEX = """\
+dev: "/path/to/config-file.yaml"
+"prod/pex_config": "s3://bucket/apdb-prod.py"
+"prod/yaml": "s3://bucket/apdb-prod.yaml"
+"""
+
+BAD_INDEX = """\
+dev: ["/path/to/config-file.yaml"]
+prod:
+    pex_config: "s3://bucket/apdb-prod.py"
+    yaml: "s3://bucket/apdb-prod.yaml"
+"""
+
+
+class ApdbIndexTestCase(unittest.TestCase):
+    """A test case for ApdbIndex class."""
+
+    def setUp(self) -> None:
+        self.tempdir = tempfile.mkdtemp()
+        self.valid_index_path = os.path.join(self.tempdir, "apdb-index.yaml")
+        with open(self.valid_index_path, "w") as index_file:
+            index_file.write(VALID_INDEX)
+        self.bad_index_path = os.path.join(self.tempdir, "bad-index.yaml")
+        with open(self.bad_index_path, "w") as index_file:
+            index_file.write(BAD_INDEX)
+        self.missing_index_path = os.path.join(self.tempdir, "missing-index.yaml")
+
+    def tearDown(self) -> None:
+        shutil.rmtree(self.tempdir, ignore_errors=True)
+
+    def test_get_apdb_uri(self) -> None:
+        """Check get_apdb_uri against existing valid file."""
+        index = ApdbIndex(self.valid_index_path)
+        self.assertEqual(index.get_apdb_uri("dev"), ResourcePath("/path/to/config-file.yaml"))
+        self.assertEqual(index.get_apdb_uri("dev", "yaml"), ResourcePath("/path/to/config-file.yaml"))
+        self.assertEqual(index.get_apdb_uri("dev", "anything"), ResourcePath("/path/to/config-file.yaml"))
+        self.assertEqual(index.get_apdb_uri("prod/yaml"), ResourcePath("s3://bucket/apdb-prod.yaml"))
+        self.assertEqual(index.get_apdb_uri("prod/pex_config"), ResourcePath("s3://bucket/apdb-prod.py"))
+        with self.assertRaises(KeyError):
+            index.get_apdb_uri("prod")
+        with self.assertRaises(KeyError):
+            index.get_apdb_uri("prod", "anything")
+        with self.assertRaises(KeyError):
+            index.get_apdb_uri("anything")
+
+    def test_missing_config(self) -> None:
+        """Check get_apdb_uri when index file is missing."""
+        index = ApdbIndex(self.missing_index_path)
+        with self.assertRaises(FileNotFoundError):
+            index.get_apdb_uri("prod")
+
+    def test_bad_config(self) -> None:
+        """Check get_apdb_uri when badly formatted YAML file."""
+        index = ApdbIndex(self.bad_index_path)
+        with self.assertRaises(TypeError):
+            index.get_apdb_uri("prod")
+
+    def test_get_known_labels(self) -> None:
+        """Check get_known_labels."""
+        index = ApdbIndex(self.valid_index_path)
+        self.assertEqual(index.get_known_labels(), {"dev", "prod/yaml", "prod/pex_config"})
+
+        index = ApdbIndex(self.bad_index_path)
+        self.assertEqual(index.get_known_labels(), set())
+
+        index = ApdbIndex(self.missing_index_path)
+        self.assertEqual(index.get_known_labels(), set())
+
+    def test_get_entries(self) -> None:
+        """Check get_entries."""
```

Comment:
Assuming the implementation stays the same, there should probably be a test that neither `get_known_labels` nor `get_entries` pollute the cache with an invalid mapping (i.e., a subsequent call to `get_apdb_uri` will still catch the problem).